client/client.go

```diff
@@ -1221,21 +1233,51 @@ func (c *Client) runAllocs(update *allocUpdates) {
 			c.logger.Printf("[ERR] client: failed to update alloc '%s': %v",
 				update.exist.ID, err)
 		}
+
+		// See if the updated alloc is getting migrated
+		c.migratingAllocsLock.Lock()
+		ch, ok := c.migratingAllocs[update.updated.ID]
+		c.migratingAllocsLock.Unlock()
+		if ok {
+			// Stopping the migration if the allocation doesn't need any
+			// migration
+			if update.updated.StopMigration() {
+				close(ch)
+			}
+		}
 	}
 
 	// Start the new allocations
 	for _, add := range diff.added {
 		// If the allocation is chanined and the previous allocation hasn't
 		// terminated yet, then add the alloc to the blocked queue.
-		if ar, ok := c.getAllocRunners()[add.PreviousAllocation]; ok && !ar.Alloc().Terminated() {
+		ar, ok := c.getAllocRunners()[add.PreviousAllocation]
+		if ok && !ar.Alloc().Terminated() {
 			c.logger.Printf("[DEBUG] client: added alloc %q to blocked queue", add.ID)
 			c.blockedAllocsLock.Lock()
 			c.blockedAllocations[add.PreviousAllocation] = add
 			c.blockedAllocsLock.Unlock()
 			continue
 		}
 
-		if err := c.addAlloc(add); err != nil {
+		// This means the allocation has a previous allocation on another node
+		// so we will block for the previous allocation to complete
+		if add.PreviousAllocation != "" && !ok {
```

Comment:
@dadgar No, since we have decided to always block an allocation before the previous one goes to the terminal state in the case it's an update we will block the allocation. The block logic would decide whether to migrate or not based on the configuration of the `EphemeralDisk`.